utility_analysis_new/combiners.py

```diff
@@ -215,3 +215,130 @@ def metrics_names(self) -> List[str]:
 
     def explain_computation(self):
         pass
+
+
+@dataclass
+class SumUtilityAnalysisMetrics:
+    """Stores metrics for the sum utility analysis.
+
+    Args:
+        sum: actual sum of contributions per partition.
+        per_partition_contribution_error: the amount of sum contribution error in a partition.
+        expected_cross_partition_error: the estimated amount of error across partitions.
+        std_cross_partition_error: the standard deviation of the contribution bounding error.
+        std_noise: the noise standard deviation for DP count.
+        noise_kind: the type of noise used.
+    """
+    sum: float
+    per_partition_contribution_error: float
+    expected_cross_partition_error: float
+    std_cross_partition_error: float
+    std_noise: float
+    noise_kind: pipeline_dp.NoiseKind
+
+
+@dataclass
+class UtilityAnalysisSumAccumulator:
+    sum: float
+    per_partition_contribution_error: float
+    expected_cross_partition_error: float
+    var_cross_partition_error: float
+    partition_selection_accumulator: Optional[PartitionSelectionAccumulator]
+
+    def __add__(self, other):
+        ps_accumulator = None
+        if self.partition_selection_accumulator is not None:
+            ps_accumulator = self.partition_selection_accumulator + other.partition_selection_accumulator
+
+        return UtilityAnalysisSumAccumulator(
+            self.sum + other.sum, self.per_partition_contribution_error +
+            other.per_partition_contribution_error,
+            self.expected_cross_partition_error +
+            other.expected_cross_partition_error,
+            self.var_cross_partition_error + other.var_cross_partition_error,
+            ps_accumulator)
+
+
+class UtilityAnalysisSumCombiner(pipeline_dp.Combiner):
+    """A combiner for utility analysis sums."""
+    AccumulatorType = UtilityAnalysisSumAccumulator
+
+    def __init__(self, params: pipeline_dp.combiners.CombinerParams,
+                 is_public_partitions: bool):
+        self._params = params
+        self._is_public_partitions = is_public_partitions
+
+    def create_accumulator(self, data: Tuple[Sized, int]) -> AccumulatorType:
+        """Create an accumulator for data.
+
+        Args:
+            data is a Tuple containing; 1) a list of the user's contributions for a single partition, and 2) the total
+            number of partitions a user contributed to.
+
+        Returns:
+            An accumulator with the sum of contributions and the contribution error.
+        """
+        if not data or not len(data[0]):
+            return UtilityAnalysisSumAccumulator(0, 0, 0, 0, None)
+        values, n_partitions = data
+        max_partitions = self._params.aggregate_params.max_partitions_contributed
+        prob_keep_partition = min(1, max_partitions /
+                                  n_partitions) if n_partitions > 0 else 0
+        partition_sum = np.sum(values)
```

Comment:
@dvadym I'm curious about an issue with this line which gives rounding error sometimes. I tried `math.fsum` which is reported to sometimes fix the issue, but it is still a problem even with 2 small numbers in the simple tests. Do we have a precision standard so it could be rounded? I think the default is 17 digits. Would you suggest using `Decimal` type or are we OK with some rounding?

Reply:
Thanks for this, it works!